File: rust/src/segments.rs
use std::rc::Rc;

use web_sys::HtmlImageElement;

use crate::{engine::{Image, Point, SpriteSheet}, obstacles::{Platform, FLOATING_PLATFORM_SPRITES, FLOATING_PLATFORM_BOUNDING_BOXES, Obstacle, Barrier, LOW_PLATFORM, FIRST_PLATFORM, HIGH_PLATFORM}};

fn create_floating_platform(sprite_sheet: Rc<SpriteSheet>, position: Point) -> Platform {
    Platform::new(
        sprite_sheet,
        position,
        &FLOATING_PLATFORM_SPRITES,
        &FLOATING_PLATFORM_BOUNDING_BOXES,
    )
}

pub fn stone_and_platform(
    stone: HtmlImageElement,
    sprite_sheet: Rc<SpriteSheet>,
    offset_x: i16,
) -> Vec<Box<dyn Obstacle>> {
    const INITIAL_STONE_OFFSET: i16 = 150;
    const STONE_ON_GROUND: i16 = 520;
    vec![
        Box::new(Barrier::new(Image::new(
            stone,
            Point {
                x: offset_x + INITIAL_STONE_OFFSET,
                y: STONE_ON_GROUND,
            },
        ))),
        Box::new(create_floating_platform(
            sprite_sheet,
            Point {
                x: offset_x + FIRST_PLATFORM,
                y: LOW_PLATFORM,
            },
        )),
    ]
}

pub fn platform_and_stone(
    stone: HtmlImageElement,
    sprite_sheet: Rc<SpriteSheet>,
    offset_x: i16,
) -> Vec<Box<dyn Obstacle>> {
    const INITIAL_STONE_OFFSET: i16 = 150;
    const STONE_ON_GROUND: i16 = 520;
    vec![
        Box::new(Barrier::new(Image::new(
            stone,
            Point {
                x: offset_x + FIRST_PLATFORM,
                y: STONE_ON_GROUND,
            },
        ))),
        Box::new(create_floating_platform(
            sprite_sheet,
            Point {
                x: offset_x + INITIAL_STONE_OFFSET,
                y: HIGH_PLATFORM,
            },
        )),
    ]
}

File: js/src/segments.ts
import { IPoint, SpriteSheet, Image } from "./engine";
import {
  Barrier,
  FIRST_PLATFORM,
  FLOATING_PLATFORM_BOUNDING_BOXES,
  FLOATING_PLATFORM_SPRITES,
  HIGH_PLATFORM,
  IObstacle,
  LOW_PLATFORM,
  Platform,
} from "./obstacles";

export function create_floating_platform(sprite_sheet: SpriteSheet, position: IPoint): Platform {
  return new Platform(sprite_sheet, position, FLOATING_PLATFORM_SPRITES, FLOATING_PLATFORM_BOUNDING_BOXES);
}

export function stone_and_platform(
  stone: HTMLImageElement,
  sprite_sheet: SpriteSheet,
  offset_x: number
): Array<IObstacle> {
  const INITIAL_STONE_OFFSET: number = 150;
  const STONE_ON_GROUND: number = 520;
  return [
    new Barrier(
      new Image(stone, {
        x: offset_x + INITIAL_STONE_OFFSET,
        y: STONE_ON_GROUND,
      })
    ),
    create_floating_platform(sprite_sheet, {
      x: offset_x + FIRST_PLATFORM,
      y: LOW_PLATFORM,
    }),
  ];
}

export function platform_and_stone(
  stone: HTMLImageElement,
  sprite_sheet: SpriteSheet,
  offset_x: number
): Array<IObstacle> {
  const INITIAL_STONE_OFFSET: number = 150;
  const STONE_ON_GROUND: number = 520;
  return [
    new Barrier(
      new Image(stone, {
        x: offset_x + FIRST_PLATFORM,
        y: STONE_ON_GROUND,
      })
    ),
    create_floating_platform(sprite_sheet, {
      x: offset_x + INITIAL_STONE_OFFSET,
      y: HIGH_PLATFORM,
    }),
  ];
}
